server/map_rpc_server.go

```diff
@@ -697,14 +674,36 @@ func (t *TrillianMapServer) snapshotForTree(ctx context.Context, tree *trillian.
 	return tx, err
 }
 
-// hasDuplicates returns an error if there are duplicates in indexes.
-func hasDuplicates(indexes [][]byte) error {
-	set := make(map[string]bool)
-	for _, i := range indexes {
-		if set[string(i)] {
-			return status.Errorf(codes.InvalidArgument, "index %x requested more than once", i)
-		}
-		set[string(i)] = true
+// requestIndexValidator validates the indices in a single map request.
+type requestIndexValidator struct {
+	// The index size in bytes
+	indexSize   int
+	seenIndices map[string]bool
+}
+
+func newRequestIndexValidator(indexSize int) *requestIndexValidator {
+	return &requestIndexValidator{
+		indexSize:   indexSize,
+		seenIndices: make(map[string]bool),
+	}
+}
+
+func (c *requestIndexValidator) validate(index []byte) error {
+	// The parameter is named 'index' (here and in the RPC API) because it's the ordinal number
+	// of the leaf, but that number is obtained by hashing the key value that corresponds to the
```

Comment:
Probably out of this PR's scope, but OTOH maybe a follow-up PR could start a little naming revolution:

The notion of "hashes of keys" feels confusing, and mixes abstractions a bit. We could just say that the map tree has a fixed-length keyspace, and then it's up to the client/personality to satisfy this limitation (e.g. by hashing the "real" keys, zero-padding them, or doing magic like 248bytes(hash)+8bytes(local_index)). The advantage: this allows throwing away "index" (well, maybe not if it's explicitly used as "leaf index") and "hash of a key" terms, in favor of just "key".

Reply:
A future PR can introduce a consistent happy naming scheme here. It's beyond the scope of this small refactor. This comment has been moved down from above.